src/algorithm/navigation_algorithm.py

```diff
@@ -0,0 +1,184 @@
+import open3d as o3d
+import numpy as np
+import time
+from queue import PriorityQueue
+import random
+import heapq
+import matplotlib.pyplot as plt
+
+
+class AStarGrid:
+    def __init__(self, rows, cols, start_x, start_y, end_x, end_y):
+        self.rows = rows
+        self.cols = cols
+        self.spacing = 1.0
+        self.grid = np.zeros((rows, cols), dtype=int)
+        self.start = (start_x, start_y)
+        self.goal = (end_x, end_y)
+
+    def generate_random_grid(self, rows, cols):
+        # Assign random costs to each cell
+        for i in range(rows):
+            for j in range(cols):
+                if (i, j) != self.start and (i, j) != self.goal:
+                    k = random.randint(0, 6)
+                    if k > 1:
+                        self.grid[i, j] = 1
+                    else:
+                        self.grid[i, j] = 0
+                else:
+                    self.grid[i, j] = 1
+
+    def a_star_search(self):
+        start_node = self.start
+        goal_node = self.goal
+
+        frontier = []
+        heapq.heappush(frontier, (0, start_node))
+
+        came_from = {start_node: None}
+        cost_so_far = {start_node: 0}
+
+        visited = set()
+
+        while frontier:
+            _, current = heapq.heappop(frontier)
+            visited.add(current)
+
+            if current == goal_node:
+                break
+
+            for next_node in self.get_neighbors(current):
+                if next_node in visited or self.grid[next_node] == 0:
+                    continue
+                else:
+                    new_cost = cost_so_far[current] + self.cost(current, next_node)
+                    if next_node not in cost_so_far and new_cost == cost_so_far[current] + 1:
+                        cost_so_far[next_node] = new_cost
+                        priority = new_cost + self.heuristic(goal_node, next_node)
+                        heapq.heappush(frontier, (priority, next_node))
+                        came_from[next_node] = current
+
+        path = []
+        current = goal_node
+        while current != start_node:
+            path.append(current)
+            current = came_from[current]
+        path.append(start_node)
+        path.reverse()
+        print(cost_so_far)
+        return path
+
+    def get_neighbors(self, node):
+        neighbors = []
+        row, col = node
+        directions = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
```

Comment:
on chodzi na ukos ?? jeśli tak to super tylko z tego co pamiętam to tego nie robił bo jeśli nie robi to chyba tylko 4 wystarczą 